python-dice-roll/source_code_final/dice.py

```diff
@@ -0,0 +1,189 @@
+    For example, if `dice_values = [4, 1, 3, 2]` then the string
+    returned looks like this:
+
+    ~~~~~~~~~~~~~~~~~~~ RESULTS ~~~~~~~~~~~~~~~~~~~
+    ┌─────────┐ ┌─────────┐ ┌─────────┐ ┌─────────┐
+    │  ●   ●  │ │         │ │  ●      │ │  ●      │
+    │         │ │    ●    │ │    ●    │ │         │
+    │  ●   ●  │ │         │ │      ●  │ │      ●  │
+    └─────────┘ └─────────┘ └─────────┘ └─────────┘
+    """
+    # Generate a list of dice faces from DICE_ART
+    dice_faces = []
+    for value in dice_values:
+        dice_faces.append(DICE_ART[value])
+
+    # Generate a list containing the dice faces rows
+    dice_faces_rows = []
+    for row_idx in range(DIE_HEIGHT):
+        row_components = []
+        for die in dice_faces:
+            row_components.append(die[row_idx])
+        row_string = DIE_FACE_SEPARATOR.join(row_components)
+        dice_faces_rows.append(row_string)
+
+    # Generate header with the word "RESULTS" centered
+    width = len(dice_faces_rows[0])
+    diagram_header = " RESULTS ".center(width, "~")
+
+    dice_faces_diagram = "\n".join([diagram_header] + dice_faces_rows)
+    return dice_faces_diagram
+
+
+# EXERCISE:
+# Refactor generate_dice_faces_diagram() to satisfy the single-responsibility
+# principle. Take advantage of _get_dice_faces() and write another helper
+# function called _generate_dice_faces_rows() to extract the functionality
+# from lines 115 to 121.
+
+# def generate_dice_faces_diagram(dice_values):
+#     """Return ASCII die faces diagram for `dice_values`.
+
+#     The string returned contains an ASCII representation of each die.
+#     For example, if `dice_values = [4, 1, 3, 2]` then the string
+#     returned looks like this:
+
+#     ~~~~~~~~~~~~~~~~~~~ RESULTS ~~~~~~~~~~~~~~~~~~~
+#     ┌─────────┐ ┌─────────┐ ┌─────────┐ ┌─────────┐
+#     │  ●   ●  │ │         │ │  ●      │ │  ●      │
+#     │         │ │    ●    │ │    ●    │ │         │
+#     │  ●   ●  │ │         │ │      ●  │ │      ●  │
+#     └─────────┘ └─────────┘ └─────────┘ └─────────┘
+#     """
+#     dice_faces = _get_dice_faces(dice_values)
+#     dice_faces_rows = _generate_dice_faces_rows(dice_faces)
+
+#     # Generate header with the word "RESULTS" centered
+#     width = len(dice_faces_rows[0])
+#     diagram_header = " RESULTS ".center(width, "~")
+
+#     dice_faces_diagram = "\n".join([diagram_header] + dice_faces_rows)
+#     return dice_faces_diagram
+
+
+# def _get_dice_faces(dice_values):
+#     dice_faces = []
+#     for value in dice_values:
+#         dice_faces.append(DICE_ART[value])
+#     return dice_faces
+
+
+# def _generate_dice_faces_rows(dice_faces):
+#     dice_faces_rows = []
+#     for row_idx in range(DIE_HEIGHT):
+#         row_components = []
+#         for die in dice_faces:
+#             row_components.append(die[row_idx])
+#         row_string = DIE_FACE_SEPARATOR.join(row_components)
+#         dice_faces_rows.append(row_string)
+#     return dice_faces_rows
```

Comment:
I think it's awkward to have this commented-out code in the source code, especially in so many different steps.

IMO this could get resolved if you move the refactoring into its own step (Step 4) instead of having it as an optional exercise in step 3.

I've mentioned this in more detail over in the tutorial-drafts PR: https://github.com/realpython/tutorial-drafts/pull/744/files#diff-1a870bf5e170073f6e7784c862223407c720d270c35f8d181d2e14cbe655cf9aR398, but I can see immediate advantages of going down that road in how it'd relate to the code in the source-code folders here.